nextest-runner/src/test_output.rs

```diff
@@ -0,0 +1,258 @@
+    }
+
+    /// Gets the raw stderr buffer
+    #[inline]
+    pub fn stderr(&self) -> Bytes {
+        self.as_buf(false)
+    }
+
+    fn as_buf(&self, stdout: bool) -> Bytes {
+        let count = self
+            .chunks
+            .iter()
+            .filter_map(|oc| (oc.stdout == stdout).then_some(oc.range.len()))
+            .sum();
+
+        self.chunks
+            .iter()
+            .fold(bytes::BytesMut::with_capacity(count), |mut acc, oc| {
+                if oc.stdout != stdout {
+                    return acc;
+                }
+
+                acc.extend_from_slice(&self.buf[oc.range.clone()]);
+                acc
+            })
+            .freeze()
+    }
+}
+
+/// Captures the stdout and/or stderr streams into a buffer, indexed on each
+/// chunk of output including the timestamp and which stream it came from
+pub struct TestOutputAccumulator {
+    buf: BytesMut,
+    chunks: Vec<OutputChunk>,
+    start: Instant,
+}
+
+impl TestOutputAccumulator {
+    /// Creates a new test accumulator to capture output from a child process
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            buf: BytesMut::with_capacity(4 * 1024),
+            chunks: Vec::new(),
+            start: Instant::now(),
+        }
+    }
+
+    /// Similar to [`bytes::BytesMut::freeze`], this is called when output
+    /// capturing is complete to create a [`TestOutput`] of the complete
+    /// captured output
+    pub fn freeze(self) -> TestOutput {
+        TestOutput {
+            buf: self.buf.freeze(),
+            chunks: self.chunks,
+            start: self.start,
+        }
+    }
+
+    /// Gets a writer the can be used to write to the accumulator as if a child
+    /// process was writing to stdout
+    #[inline]
+    pub fn stdout(&mut self) -> TestOutputWriter<'_> {
+        TestOutputWriter {
+            acc: self,
+            stdout: true,
+        }
+    }
+
+    /// Gets a writer the can be used to write to the accumulator as if a child
+    /// process was writing to stderr
+    #[inline]
+    pub fn stderr(&mut self) -> TestOutputWriter<'_> {
+        TestOutputWriter {
+            acc: self,
+            stdout: false,
+        }
+    }
+}
+
+/// Provides [`std::io::Write`] and [`std::fmt::Write`] implementations for a
+/// single stream of a [`TestOutputAccumulator`]
+pub struct TestOutputWriter<'acc> {
+    acc: &'acc mut TestOutputAccumulator,
+    stdout: bool,
+}
+
+impl<'acc> std::io::Write for TestOutputWriter<'acc> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let start = self.acc.buf.len();
+        self.acc.buf.extend_from_slice(buf);
+        self.acc.chunks.push(OutputChunk {
+            range: start..self.acc.buf.len(),
+            timestamp: Instant::now(),
+            stdout: self.stdout,
+        });
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
```

Comment:
Just guessing this is related to a bug I noticed: When I run multiple integration tests at least, only the _last_ module in those tests is present in the stdout of the report. Snippet of an example run, so I know tests are run, just not reported i the json:

```
       FAIL [   0.003s] pba-qualifier-exam::h_advanced_traits_tests ic_4
        FAIL [   0.002s] pba-qualifier-exam::h_advanced_traits_tests mixed_2
        FAIL [   0.003s] pba-qualifier-exam::h_advanced_traits_tests ic_3
        FAIL [   0.003s] pba-qualifier-exam::h_advanced_traits_tests ic_2
        FAIL [   0.003s] pba-qualifier-exam::h_advanced_traits_tests omni_3
        FAIL [   0.004s] pba-qualifier-exam::i_extension_traits_tests simple_non_empty
        FAIL [   0.004s] pba-qualifier-exam::i_extension_traits_tests simple_empty
        FAIL [   0.003s] pba-qualifier-exam::i_extension_traits_tests trait_empty
        FAIL [   0.003s] pba-qualifier-exam::i_extension_traits_tests trait_non_empty
        FAIL [   0.002s] pba-qualifier-exam::k_macros_tests map_single_entry
        PASS [   0.003s] pba-qualifier-exam::k_macros_tests map_empty
        FAIL [   0.003s] pba-qualifier-exam::k_macros_tests map_three_entries
        FAIL [   0.003s] pba-qualifier-exam::m_builder_tests builder_all_set
        FAIL [   0.003s] pba-qualifier-exam::m_builder_tests builder_correct_defaults
        FAIL [   0.002s] pba-qualifier-exam::m_builder_tests builder_fail
        FAIL [   0.002s] pba-qualifier-exam::m_builder_tests typed_builder_correct_defaults
        FAIL [   0.002s] pba-qualifier-exam::m_builder_tests typed_builder_all_set
        FAIL [   0.372s] pba-qualifier-exam::h_advanced_traits_tests british_engine_only_works_with_btu
        FAIL [   0.818s] pba-qualifier-exam::h_advanced_traits_tests internal_combustion_only_works_with_uranium
        FAIL [   1.318s] pba-qualifier-exam::h_advanced_traits_tests nuclear_reactor_only_works_with_uranium
        FAIL [   1.741s] pba-qualifier-exam::k_macros_tests impl_get_single_pub_u64
        FAIL [   2.240s] pba-qualifier-exam::k_macros_tests impl_get_mixed_types_and_vis
        FAIL [   2.749s] pba-qualifier-exam::k_macros_tests impl_get_single_u32
        FAIL [   3.758s] pba-qualifier-exam::k_macros_tests impl_get_multiple_arrays
     Running [ 00:00:03] [=========================================================================================================> ] 101/102: 0 running, 4 passed, 98 failed, 0 skipped     {"type":"suite","event":"started","test_count":7,"nextest":{"crate":"pba-qualifier-exam","test_binary":"k_macros_tests","kind":"test"}}
{"type":"test","event":"started","name":"pba-qualifier-exam::k_macros_tests$impl_get_mixed_types_and_vis"}
{"type":"test","event":"started","name":"pba-qualifier-exam::k_macros_tests$impl_get_multiple_arrays"}
{"type":"test","event":"started","name":"pba-qualifier-exam::k_macros_tests$impl_get_single_pub_u64"}
{"type":"test","event":"started","name":"pba-qualifier-exam::k_macros_tests$impl_get_single_u32"}
{"type":"test","event":"started","name":"pba-qualifier-exam::k_macros_tests$map_empty"}
{"type":"test","event":"started","name":"pba-qualifier-exam::k_macros_tests$map_single_entry"}
{"type":"test","event":"started","name":"pba-qualifier-exam::k_macros_tests$map_three_entries"}
{"type":"test","event":"failed","name":"pba-qualifier-exam::k_macros_tests$map_single_entry","exec_time":0.00232729,"stdout":"\nrunning 1 test\nthread '<unnamed>' panicked at tests/k_macros_tests.rs:19:5:\nassertion `left == right` failed\n  left: None\n right: Some(3)\nnote: run with `RUST_BACKTRACE=1` environment variable to display a backtrace\nthread 'map_single_entry' panicked at tests/k_macros_tests.rs:15:1:\nexplicit panic\ntest map_single_entry ... FAILED\n\nfailures:\n\nfailures:\n    map_single_entry\n\ntest result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 6 filtered out; finished in 0.00s\n\n"}
```

Reply:
That flush is irrelevant, that's only a writer for an in-memory buffer, it has nothing to do with stdout.

Reply:
That being said I can repro the bug and will fix it.

Reply:
Fixed.